### DISCUSSION TITLE
PorousFlow: P&T outside range-non linear solver does not converge

### DISCUSSION THREAD

**Discussion by rksin8** [idaholab/moose#19312] [Q&A Modules: General] [answered]
Hi,
I am working on CO2 injection in a brine saturated reservoir problem. The model runs for a few time steps after that it message pressure & temperature outside the tabulated range and it does not converge. 

![image](https://user-images.githubusercontent.com/22324848/140675508-72ba379a-e160-4749-afbd-022b4370a100.png)

I tried with different capillary and relative permeability relation, mesh refinement, different tabulated range, injection rate as given [(nonlinear_convergence_problems)](https://mooseframework.inl.gov/modules/porous_flow/nonlinear_convergence_problems.html) here but still no improvement.  I used the same parameters in [Dumux](https://dumux.org/) and it works fine. 
![image](https://user-images.githubusercontent.com/22324848/140674730-5f841f2f-5f0b-4b76-ad6e-8c8b89cc8e5d.png)

My input file: https://github.com/rksin8/co2-bine/blob/main/co2_brine_theis.i
Could anyone please let me know what causes this error and how to fix it?

Thanks in advance!

**Comment by cpgr** [comment, marked as answer]
I have to head out for a bit now, but will take a look tonight.

**Reply by cpgr** [reply]
In my experience, these issues are mainly caused by something happening at the boundaries. I had a look, and the salt profile at the rightmost boundary was doing something funny. I changed that BC to a simple DirichletBC
```
  [right_xnacl]
    type = DirichletBC
    variable = xnacl
    boundary = 'right'
    value = 0.1
  []
```

(and deleted the `gravity = ` param in each kernel block, which weren't all the same! - the gravity value in the global params block is then used consistently), and the problem ran as expected (note: I did coarsen the mesh to make it a bit faster). 

![co2inj](https://user-images.githubusercontent.com/7664823/140813492-cb4d9e7c-a403-450a-917e-796f16e38622.png)

I think the injection rate might not be the same as the Dumux model, though, as this profile is after only ~15 days, so you might want to double check that.

**Reply by rksin8** [reply]
Hi Chris,
Thank you for your help. Now I am getting results very close to the dumux one.

![image](https://user-images.githubusercontent.com/22324848/140897747-09e1136b-e0e9-4d66-baee-575860e48aa0.png)

![image](https://user-images.githubusercontent.com/22324848/140897812-f11add08-b741-47bb-969c-b9db701458d3.png)

**Reply by cpgr** [reply]
Great! It looks like you might need to change the BC for the total mass fraction z at the RHS to get a better match.
